server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java

```diff
@@ -76,9 +77,9 @@ public ServerAmpleImpl(ServerContext context) {
   }
 
   @Override
-  public Ample.TabletMutator mutateTablet(KeyExtent extent) {
+  public Ample.TabletMutator mutateTablet(KeyExtent extent, ServiceLock lock) {
```

Comment:
Not sure if this will be better.  These changes require passing the service lock around all of the time.  Was wondering if we could instead pass the service lock in when creating ServerAmpleImpl.  Maybe this would not require passing the service lock to so many methods.

Was playing around w/ the idea the when the ServerContext is created a supplier is passed in the for the service lock.  The idea being that server processes could set it later when its available.  Something like AtomicReference.get could be used as the impl for the supplier.  The tserver creates the AtomicRef and passes it to the ServerContext and later when it creates the ServiceLock it sets the atomic ref.

```diff
diff --git a/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java b/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
index e58645d00b..7ca94fd785 100644
--- a/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
+++ b/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
@@ -103,12 +103,13 @@ public class ServerContext extends ClientContext {
   private final Supplier<AuditedSecurityOperation> securityOperation;
   private final Supplier<CryptoServiceFactory> cryptoFactorySupplier;
   private final Supplier<LowMemoryDetector> lowMemoryDetector;
+  private final Supplier<ServiceLock> serviceLockSupplier;
 
-  public ServerContext(SiteConfiguration siteConfig) {
-    this(new ServerInfo(siteConfig));
+  public ServerContext(SiteConfiguration siteConfig, Supplier<ServiceLock> serviceLockSupplier) {
+    this(new ServerInfo(siteConfig), serviceLockSupplier);
   }
 
-  private ServerContext(ServerInfo info) {
+  private ServerContext(ServerInfo info, Supplier<ServiceLock> serviceLockSupplier) {
     super(SingletonReservation.noop(), info, info.getSiteConfiguration(), Threads.UEH);
     this.info = info;
     zooReaderWriter = new ZooReaderWriter(info.getSiteConfiguration());
@@ -129,6 +130,7 @@ public class ServerContext extends ClientContext {
         memoize(() -> new AuditedSecurityOperation(this, SecurityOperation.getAuthorizor(this),
             SecurityOperation.getAuthenticator(this), SecurityOperation.getPermHandler(this)));
     lowMemoryDetector = memoize(() -> new LowMemoryDetector());
+    this.serviceLockSupplier = serviceLockSupplier;
   }
 
   /**
@@ -275,8 +277,13 @@ public class ServerContext extends ClientContext {
     return cryptoFactorySupplier.get();
   }
 
+  public ServiceLock getServiceLock(){
+    return serviceLockSupplier.get();
+  }
+
   @Override
   public Ample getAmple() {
+    // in the impl of ServerAmpleImpl we can call getServiceLock() as needed
     return new ServerAmpleImpl(this);
   }

```

If we had something like that, then this code could be

```java
public Ample.TabletMutator mutateTablet(KeyExtent extent) {
    TabletsMutator tmi = mutateTablets();
    Ample.TabletMutator tabletMutator = tmi.mutateTablet(extent, context.getServiceLock());
    ((TabletMutatorBase) tabletMutator).setCloseAfterMutate(tmi);
    return tabletMutator;
  }
```

Reply:
@cshannon's suggestion:
   `Looking at this, is it possible to move the passing of the lock to the construction of the mutator?`
   
 I'm fine with changing the how/where the lock is supplied. The biggest issue I'm having right now is I can't figure out why BigRootTabletIT is failing.

Reply:
Re the supplier in my initial comment, I initially thought about having a setter method instead.  Was thinking the supplier made apparent that its a mandatory requirement to provide it even though its not immediately available. A setter is something that someone could forget to call.

Reply:
Yeah, I'm not opposed to changing the plumbing

Reply:
I re-worked the service lock plumbing in 0243ff0.